lib/AST/Module.cpp

```diff
@@ -1586,6 +1594,34 @@ bool SourceFile::isImportedImplementationOnly(const ModuleDecl *module) const {
   return !imports.isImportedBy(module, getParentModule());
 }
 
+void SourceFile::lookupImportedSPIs(const ModuleDecl *importedModule,
+                                    SmallVectorImpl<Identifier> &spis) const {
+  for (auto &import : Imports) {
+    if (import.importOptions.contains(ImportFlags::SPIAccessControl) &&
+        importedModule == std::get<ModuleDecl*>(import.module)) {
+      auto importedSpis = import.spis;
+      spis.append(importedSpis.begin(), importedSpis.end());
+    }
+  }
+}
+
+bool SourceFile::isImportedAsSPI(const ValueDecl *targetDecl) const {
+  if (!targetDecl->getAttrs().hasAttribute<SPIAccessControlAttr>())
+    return false;
+
+  auto targetModule = targetDecl->getModuleContext();
+  SmallVector<Identifier, 4> importedSpis;
+  lookupImportedSPIs(targetModule, importedSpis);
+
+  for (auto attr : targetDecl->getAttrs().getAttributes<SPIAccessControlAttr>())
+    for (auto declSPI : attr->getSPINames())
+      for (auto importedSPI : importedSpis)
```

Comment:
hmm, three nested loops.. can we use `Set` somewhere to avoid iterating?

Reply:
I plan on moving the logic collecting the SPI groups declared on a decl to a request in order to cover what Slava pointed out with decls inheriting SPI from their parent. With it I can probably add a function for the previous comment too.